massa-execution/src/interface_impl.rs

```diff
@@ -27,17 +36,117 @@ impl InterfaceClone for InterfaceImpl {
 
 impl Interface for InterfaceImpl {
     fn get_module(&self, address: &String) -> Result<Vec<u8>> {
-        let bytecode = {
-            let address = Address::from_str(address)?;
-            let context = self
-                .context
-                .lock()
-                .expect("Failed to acquire lock on context.");
-            context.ledger_step.get_module(&address)
-        };
+        let address = Address::from_str(address)?;
+        let bytecode = { context_guard!(self).ledger_step.get_module(&address) };
         match bytecode {
-            Some(bytecode) => Ok(bytecode),
+            Some(bytecode) => {
+                context_guard!(self).call_stack.push_back(address);
+                Ok(bytecode)
+            }
             _ => bail!("Error bytecode not found"),
         }
     }
+
+    fn exit_success(&self) -> Result<()> {
+        match context_guard!(self).call_stack.pop_back() {
+            Some(_) => Ok(()),
+            _ => bail!("Call stack Out of bound"),
+        }
+    }
+
+    /// Requires a new address that contains the sent bytecode.
+    ///
+    /// Generate a new address with a concatenation of the block_id hash and the index of address owned in context.
+    ///
+    /// Insert in the ledger the given bytecode in the generated address
+    fn create_module(&self, module: &Bytecode) -> Result<assembly_simulator::Address> {
+        let (block_id, index) = {
+            let context = context_guard!(self);
+            (context.opt_block_id, context.owned_addresses.len())
+        };
+        let block_id = match block_id {
+            Some(block_id) => block_id,
+            _ => bail!("Failed to read current context"),
+        };
+        let mut data = block_id.to_bytes().to_vec();
+        data.append(&mut index.to_be_bytes().to_vec());
```

Comment:
non-deterministic !

owned_addresses.len() returns an usize which is u64 on 64bit systems, 32 on 32bit systems etc... let's force conversion to u64 beforehand

Reply:
FIxed, but linked with another thread ;) https://github.com/massalabs/massa/pull/2048#discussion_r776802683